# <b> Data Retrieval </b>
Data retrieval refers to the process of accessing and extracting data from a storage system, such as a database, file system, or web server. The retrieved data can then be used for various purposes, such as analysis, reporting, or presentation.

Data retrieval typically involves the use of specialized software tools, such as database management systems (DBMS) or search engines, that are designed to efficiently locate and retrieve specific pieces of information from large volumes of data. The process may involve querying a database using a structured query language (SQL), or using keyword-based search techniques to locate specific files or documents.


In [9]:
# importing the important libraries

import pandas as pd
import numpy as np

In [10]:
data = pd.read_csv('Dataset/Walmart.csv')

# <b> Basic Data Exploration and Data Cleaning </b>

Basic data exploration is the process of examining and summarizing the key characteristics of a dataset in order to gain a better understanding of its structure, patterns, and relationships. It is an essential step in any data analysis project, as it provides important insights that can inform subsequent data processing, modeling, and interpretation.

Data cleaning is an important step in any data analysis project, as it ensures that the data is reliable and suitable for analysis. It can be a time-consuming and iterative process, requiring careful attention to detail and a thorough understanding of the data and its context.


In [11]:
data.head(10)

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   
5      1  12-03-2010    1439541.59             0        57.79       2.667   
6      1  19-03-2010    1472515.79             0        54.58       2.720   
7      1  26-03-2010    1404429.92             0        51.45       2.732   
8      1  02-04-2010    1594968.28             0        62.27       2.719   
9      1  09-04-2010    1545418.53             0        65.86       2.770   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106  
5  211.380643         8.106  
6  211.215635         8.106  
7  211.018042         8.106  
8  210.820450         7.808  
9  210.622857         7.808

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   object 
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), object(1)
memory usage: 402.3+ KB


In [13]:
data.describe()

Store  Weekly_Sales  Holiday_Flag  Temperature   Fuel_Price  \
count  6435.000000  6.435000e+03   6435.000000  6435.000000  6435.000000   
mean     23.000000  1.046965e+06      0.069930    60.663782     3.358607   
std      12.988182  5.643666e+05      0.255049    18.444933     0.459020   
min       1.000000  2.099862e+05      0.000000    -2.060000     2.472000   
25%      12.000000  5.533501e+05      0.000000    47.460000     2.933000   
50%      23.000000  9.607460e+05      0.000000    62.670000     3.445000   
75%      34.000000  1.420159e+06      0.000000    74.940000     3.735000   
max      45.000000  3.818686e+06      1.000000   100.140000     4.468000   

               CPI  Unemployment  
count  6435.000000   6435.000000  
mean    171.578394      7.999151  
std      39.356712      1.875885  
min     126.064000      3.879000  
25%     131.735000      6.891000  
50%     182.616521      7.874000  
75%     212.743293      8.622000  
max     227.232807     14.313000

In [14]:
data.shape

(6435, 8)

<b> After analyzing the above dataset we can infer that the dataset has following inferences : </b>

- Store : The store number
- Date :  Week of the sales
- Weekly_Sales : Sales for the given store
- Holiday_Flag : If week is a special holiday week
- Temperature : Temperature on sales day
- Fuel_Price : Cost of fuel in the region of store
- CPI : Customer price index
- Unemployment : Prevailing unemployment rate in percentage


<b> Object 1 : </b> Handling the missing values

In [15]:
data.isnull().sum()*100/len(data)

Store           0.0
Date            0.0
Weekly_Sales    0.0
Holiday_Flag    0.0
Temperature     0.0
Fuel_Price      0.0
CPI             0.0
Unemployment    0.0
dtype: float64

<b> Inference 1 : </b> The data do not contains the null values so we can proceed with it

<b> Object 2 : </b> Converting the date column to date and time data type

In [16]:
data['Date'] = pd.to_datetime(data['Date'])

/var/folders/4r/21z41gt55tdf56zf__j7s8b80000gn/T/ipykernel_12309/1453708669.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  data['Date'] = pd.to_datetime(data['Date'])


In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         6435 non-null   int64         
 1   Date          6435 non-null   datetime64[ns]
 2   Weekly_Sales  6435 non-null   float64       
 3   Holiday_Flag  6435 non-null   int64         
 4   Temperature   6435 non-null   float64       
 5   Fuel_Price    6435 non-null   float64       
 6   CPI           6435 non-null   float64       
 7   Unemployment  6435 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(2)
memory usage: 402.3 KB


<b> Inference 2 : </b> The Date is converted to date-time data type


In [18]:
# Creating a column_list

column_list = data.columns.tolist()
print(column_list)

['Store', 'Date', 'Weekly_Sales', 'Holiday_Flag', 'Temperature', 'Fuel_Price', 'CPI', 'Unemployment']


<b> Object 3 : </b> Handling the duplicate data points

In [19]:
duplicates = data[data.duplicated()]

In [20]:
duplicates

Empty DataFrame
Columns: [Store, Date, Weekly_Sales, Holiday_Flag, Temperature, Fuel_Price, CPI, Unemployment]
Index: []

<b> Inference 3 : </b> The data do not contains any duplicate tuples so we can proceed with it.

<b> Object 4 : <b> Handling the outliers in the data


We will be using the inter-quartile range for detecting the outliers in the data

In [21]:
def detect_outlier(field):
    q1 = data[field].quantile(0.25)
    q3 = data[field].quantile(0.75)
    iqr = q3-q1

    threshold = iqr*1.5
    lb = q1-threshold
    ub = q3+threshold

    outliers = [feature for feature in data[field] if feature<lb or feature>ub]
    return outliers

In [22]:
outlier_features = {}
for i in column_list:
    outlier_features[i] = len(detect_outlier(i))*100/len(data)

In [23]:
outlier_features

{'Store': 0.0,
 'Date': 0.0,
 'Weekly_Sales': 0.5283605283605284,
 'Holiday_Flag': 6.993006993006993,
 'Temperature': 0.046620046620046623,
 'Fuel_Price': 0.0,
 'CPI': 0.0,
 'Unemployment': 7.474747474747475}

<b> Sub Inference : </b> There is the outlier values that are present in the following columns only in the following percentages :

- 'Weekly_Sales': 0.5283605283605284
- 'Holiday_Flag': 6.993006993006993
- 'Temperature': 0.046620046620046623
- 'Unemployment': 7.474747474747475

In [24]:
outliers = ['Weekly_Sales','Temperature','Unemployment']

In [25]:
data['Holiday_Flag'].value_counts()

0    5985
1     450
Name: Holiday_Flag, dtype: int64

<b> Problem : </b> We have Holiday_Flag with outlier values but its categorical feature that is why

In [26]:
# Function to return the outlier bounds

def return_bounds(field):
    q1 = data[field].quantile(0.25)
    q3 = data[field].quantile(0.75)
    iqr = q3-q1

    threshold = iqr*1.5
    lb = q1-threshold
    ub = q3+threshold

    return [ub,lb]


In [27]:
for i in outliers:
    lst = return_bounds(i)
    ub = lst[0]
    median = np.median(data[i])
    data[i] = np.where(data[i]>ub,median,data[i])

In [28]:
for i in outliers:
    lst = return_bounds(i)
    lb = lst[1]
    median = np.median(data[i])
    data[i] = np.where(data[i]<lb,median,data[i])

In [29]:
outlier_features = {}
for i in column_list:
    outlier_features[i] = len(detect_outlier(i))*100/len(data)

In [30]:
outlier_features

{'Store': 0.0,
 'Date': 0.0,
 'Weekly_Sales': 0.0,
 'Holiday_Flag': 6.993006993006993,
 'Temperature': 0.0,
 'Fuel_Price': 0.0,
 'CPI': 0.0,
 'Unemployment': 1.6472416472416473}

In [31]:
lst = return_bounds('Unemployment')
lb = lst[1]
ub = lst[0]

In [32]:
data = data[(data['Unemployment'] < ub) & (data['Unemployment'] > lb)]

<b> Infrence 4 : </b> Now our data do not contains any of the missing values and the missing values in the unemployment is too less that it can be ignored

<b> Object 5 : </b> Storing the data after cleaning the data 

In [33]:
# Final checking the data

data.isnull().sum()

Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

In [34]:
data.describe()

Store  Weekly_Sales  Holiday_Flag  Temperature   Fuel_Price  \
count  6329.000000  6.329000e+03   6329.000000  6329.000000  6329.000000   
mean     22.948017  1.033313e+06      0.070153    60.557894     3.350481   
std      12.991500  5.446067e+05      0.255425    18.373092     0.457760   
min       1.000000  2.099862e+05      0.000000     7.460000     2.472000   
25%      12.000000  5.510581e+05      0.000000    47.340000     2.921000   
50%      23.000000  9.562290e+05      0.000000    62.590000     3.428000   
75%      34.000000  1.412721e+06      0.000000    74.830000     3.722000   
max      45.000000  2.685352e+06      1.000000   100.140000     4.468000   

               CPI  Unemployment  
count  6329.000000   6329.000000  
mean    172.222769      7.766344  
std      39.364851      1.105107  
min     126.064000      4.954000  
25%     131.863129      7.007000  
50%     185.533982      7.874000  
75%     212.928631      8.454000  
max     227.232807     10.641000

In [35]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6329 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         6329 non-null   int64         
 1   Date          6329 non-null   datetime64[ns]
 2   Weekly_Sales  6329 non-null   float64       
 3   Holiday_Flag  6329 non-null   int64         
 4   Temperature   6329 non-null   float64       
 5   Fuel_Price    6329 non-null   float64       
 6   CPI           6329 non-null   float64       
 7   Unemployment  6329 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(2)
memory usage: 445.0 KB


In [36]:
data.shape

(6329, 8)

In [37]:
data[data.duplicated()]

Empty DataFrame
Columns: [Store, Date, Weekly_Sales, Holiday_Flag, Temperature, Fuel_Price, CPI, Unemployment]
Index: []

In [38]:
# Storing the data to file

data.to_csv('Dataset/Data_Cleaned.csv',index=False)